### DISCUSSION TITLE
useCallback and useMemo

### DISCUSSION THREAD

**Discussion by honaseung** [honaseung/react-deep-dive#23] [Chapter3]
In my opinion, `useState` and `useEffect` are important in React. But `useCallback` and `useMemo` aren't. I made my own project by React. I didn't use it. And I finished without those. It's only for optimazation. It means it's kind of advanced skill I think. So if you are not used to React, I recommand not using those. There are lots of things you should know in React more than those. But still you ready. Here we go.

#### useCallback
> What's the purpose of `useCallback` and `useMemo`? It's for remember it. Just like `state`. If re-rendering is happened, codes run from up to down once again, make variables and do the things. Right? At this time, React meet `function`, it will make it once again without a doubt. But you know `function` has a reference as a value then save it on the memory every rendering. It's wasting memory. This is the why `useCallback` is born. Remember it, don't make it twice unless it have to.

> Now you wonder. When is do we have to make it once again? It's same as `useEffect`. Put some `state` you want to watch in dependecies array. Then React will make it once again every denpencies changed. What are good for put in? Actually, there is no answer. You can put whatever you want. But even still, It's better to put things which used inside of `useCallback`. If you don't, the function hasn't been maden properly. and it will not work as your intend.
 So `useCallabk` is for remebering the function to prevent wasting memory from making once again. You can use it only for the function. So use can get params when you use `useCallback`.

#### useMemo
> How about `useMemo`? We don't need to talk about `useMemo` a lot. It's same as `useCallback`. Do the same role. One thing different is `useMemo` isn't for function. There is one more reference value type you know. The object. `useMemo` is for remembering the object. So it can't get params when you use `useMemo`. Also `useMemo` can remember `JSX`. Maybe `JSX` is a kind of object I guess.

### Have to use
> I forgot to say the most important things. Haha sorry... <sup><del>You have to use `useCallback` or `useMemo` to give function or object to child component.</del>/</sup> React only can compare shallowly. <sup><del>So if you don't use those, reference values are changed evertime. So child component have to be re-redered every time.</del> Every parents rendering occur children rendering.</sup>So If you use something inside of `useEffect`. But something that use some `state`? You should better wrap it `useCallback` or `useMemo` unless you want to see malfunction. React can think it's changed every time!

**Comment by chaminju927** [comment]
![image](https://github.com/honaseung/react-deep-dive/assets/118426671/2e1ec45b-adec-45c4-9af7-50fcbd7376a8)
Can you explain more in detail - How useMemo's returning object related with parameter of useMemo?

**Reply by honaseung** [reply]
You can wrap a function with `useMemo` this case it will go
`const test = useMemo(() => return (p) => console.log(p), []);`
When you call this,
`test('test');`
But what if codes are going this way
`const test = useMemo((param) => return (p) => console.log(p), []);`
When you call this,
`test('test');`
Does 'test' go to 'param'? Or 'p'? The answer is 'p'. You never can reach to 'param'. That's what I meant.
